include/smack/smack.h

```diff
@@ -124,137 +355,224 @@ void __SMACK_decls() {
+  D("function {:inline} $trunc.i64.i32(p: i64) returns (i32) {p[32:0]}");
+  D("function {:inline} $trunc.i64.i16(p: i64) returns (i16) {p[16:0]}");
+  D("function {:inline} $trunc.i64.i8(p: i64) returns (i8) {p[8:0]}");
+  D("function {:inline} $trunc.i64.i1(p: i64) returns (bv1) {if p != 0bv64 then 1bv1 else 0bv1}");
+  D("function {:inline} $trunc.i32.i16(p: i32) returns (i16) {p[16:0]}");
+  D("function {:inline} $trunc.i32.i8(p: i32) returns (i8) {p[8:0]}");
+  D("function {:inline} $trunc.i32.i1(p: i32) returns (bv1) {if p != 0bv32 then 1bv1 else 0bv1}");
+  D("function {:inline} $trunc.i16.i8(p: i16) returns (i8) {p[8:0]}");
+  D("function {:inline} $trunc.i8.i1(p: i8) returns (bv1) {if p != 0bv8 then 1bv1 else 0bv1}");
 
-  D("const unique $NULL: int;");
+  D("function {:inline} $zext.i1.i64(p: bv1) returns (i64) {if p != 0bv1 then 1bv64 else 0bv64}");
+  D("function {:inline} $zext.i1.i32(p: bv1) returns (i32) {if p != 0bv1 then 1bv32 else 0bv32}");
+  D("function {:inline} $zext.i1.i16(p: bv1) returns (i16) {if p != 0bv1 then 1bv16 else 0bv16}");
+  D("function {:inline} $zext.i1.i8(p: bv1) returns (i8) {if p != 0bv1 then 1bv8 else 0bv8}");
+  D("function {:inline} $zext.i8.i64(p: i8) returns (i64) {(0bv56)++p}");
+  D("function {:inline} $zext.i8.i32(p: i8) returns (i32) {(0bv24)++p}");
+  D("function {:inline} $zext.i8.i16(p: i8) returns (i16) {(0bv8)++p}");
+  D("function {:inline} $zext.i16.i64(p: i16) returns (i64) {(0bv48)++p}");
+  D("function {:inline} $zext.i16.i32(p: i16) returns (i32) {(0bv16)++p}");
+  D("function {:inline} $zext.i32.i64(p: i32) returns (i64) {(0bv32)++p}");
+
+  D("function {:inline} $sext.i1.i64(p: bv1) returns (i64) {if p != 0bv1 then 1bv64 else 0bv64}");
+  D("function {:inline} $sext.i1.i32(p: bv1) returns (i32) {if p != 0bv1  then 1bv32 else 0bv32}");
+  D("function {:inline} $sext.i1.i16(p: bv1) returns (i16) {if p != 0bv1 then 1bv16 else 0bv16}");
+  D("function {:inline} $sext.i1.i8(p: bv1) returns (i8) {if p != 0bv1 then 1bv8 else 0bv8}");
+  D("function {:inline} $sext.i8.i64(p: i8) returns (i64) {if $sge.i8(p, 0bv8) then $zext.i8.i64(p) else (($neg.i64(1bv64))[64:8])++p}");
+  D("function {:inline} $sext.i8.i32(p: i8) returns (i32) {if $sge.i8(p, 0bv8) then $zext.i8.i32(p) else (($neg.i32(1bv32))[32:8])++p}");
+  D("function {:inline} $sext.i8.i16(p: i8) returns (i16) {if $sge.i8(p, 0bv8) then $zext.i8.i16(p) else $neg.i8(1bv8)++p}");
+  D("function {:inline} $sext.i16.i64(p: i16) returns (i64) {if $sge.i16(p, 0bv16) then $zext.i16.i64(p) else $neg.i64(1bv64)[64:16]++p}");
+  D("function {:inline} $sext.i16.i32(p: i16) returns (i32) {if $sge.i16(p, 0bv16) then $zext.i16.i32(p) else $neg.i16(1bv16)++p}");
+  D("function {:inline} $sext.i32.i64(p: i32) returns (i64) {if $sge.i32(p, 0bv32) then $zext.i32.i64(p) else $neg.i32(1bv32)++p}");
+
+  D("function {:inline} $p2i(p: ref) returns (size) {p}");
```

Comment:
LLVM's `ptrtoint` and `inttoptr` allow arbitrary integer types, and performs truncation or zero extension as appropriate. We must handle this correctly.